File: web_version/pmugg_files/guidelines/collection.js
ms.collection = function(stats, type) {
	var properties = {};
	properties[type] = ms.unorderedAlternatives.enabled ? new ms.unorderedAlternatives(type) : new ms.alternativeArray(type);
	this.node = new ms.node(this, stats, 'collection', properties);
};

ms.collection.prototype.getNode = function() {
	return this.node;
};

ms.collection.prototype.getVertices = function() {
	if (this.node.isDestroyed()) {
		return [];
	} else {
		return this.node.get('vertex');
	}
};

ms.collection.prototype.getFaces = function() {
	if (this.node.isDestroyed()) {
		return [];
	} else {
		return this.node.get('face');
	}
};

ms.collection.prototype.getRingInstances = function() {
	if (this.node.isDestroyed()) {
		return [];
	} else {
		return this.node.get('ringInstance');
	}
};

File: cpp_version/pmugg/guidelines/collection.cpp
#include "collection.h"

namespace ms {

Collection::Collection(const Stats& stats, const std::string& type) {
    std::map<std::string, std::unique_ptr<AlternativeBase>> properties;
    
    if (globalSettings.getBool("unorderedAlternatives.enabled")) {
        properties[type] = std::make_unique<UnorderedAlternatives>(type);
    } else {
        properties[type] = std::make_unique<AlternativeArray>(type);
    }
    
    node = std::make_unique<Node>(this, stats, "collection", std::move(properties));
}

Node* Collection::getNode() const {
    return node.get();
}

std::vector<Vertex*> Collection::getVertices() const {
    if (node->isDestroyed()) {
        return {};
    }
    return node->get("vertex");
}

std::vector<Face*> Collection::getFaces() const {
    if (node->isDestroyed()) {
        return {};
    }
    return node->get("face");
}

std::vector<RingInstance*> Collection::getRingInstances() const {
    if (node->isDestroyed()) {
        return {};
    }
    return node->get("ringInstance");
}

} // namespace ms 